### PR TITLE
[onert] Optimize PermuteLayer of controlflow by using enqueue functions

### PR DESCRIPTION
**Pull request by ragmani**
For issue #4414
Draft PR #4395

This commit Optimize PermuteLayer of controlflow by using enqueue functions instead of calling map().

Signed-off-by: ragmani <ragmani0216@gmail.com>

### CODE REVIEW

**Comment by ragmani** [comment]
@nnfw-bot test all

**Comment by ragmani** [comment]
@nnfw-bot test x64-builder

**Review by hseok-oh — DISMISSED** [review]

File: runtime/onert/core/include/backend/ITensor.h [hunk 1]
```diff
@@ -103,6 +103,17 @@ class ITensor
    * @note  Higer dimension will be placed on front.
    */
   virtual ir::Shape getShape() const;
+
+  virtual bool is_subtensor() const { return false; }
+  virtual bool hasClBuffer() const { return false; }
+  virtual void enqueueWriteBuffer(const void *)
```
Comment (hyunsik-yoon) [inline runtime/onert/core/include/backend/ITensor.h:109]: Shouldn't it better for ITensor to have names neutral to specific backend?
Or how about something like the following?

- hasClBuffer -> needMemoryMap: // some backend needs CPU memory to be mapped to, e.g., GPU memory or NPU memory.
- enqueueWriteBuffer -> mapMemory

Comment (ragmani) [reply runtime/onert/core/include/backend/ITensor.h:109]: > - hasClBuffer -> needMemoryMap: // some backend needs CPU memory to be mapped to, e.g., GPU memory or NPU memory.

Okay. I will.

> - enqueueWriteBuffer -> mapMemory

This method is not to map memory. How about changing it to writeBuffer?

Comment (hyunsik-yoon) [reply runtime/onert/core/include/backend/ITensor.h:109]: how about writeMappedMemory ?

Comment (ragmani) [reply runtime/onert/core/include/backend/ITensor.h:109]: This method is not related to mapping data. It writes a buffer to device memory of the backend tensor from host memory.

Comment (hyunsik-yoon) [reply runtime/onert/core/include/backend/ITensor.h:109]: I see. The concern I have is that if the name is _too general_ like `writeBuffer`, backend developer may be curious why CPU tensor cannot support `writeBuffer`. In that sense I like `enqueueWriteBuffer` (your first naming) would be better than `writeBuffer`. Or some long name like `writeBufferToDevice`(?) would also work.

Comment (ragmani) [reply runtime/onert/core/include/backend/ITensor.h:109]: I think the original(enqueueWriteBuffer) looks the best among them.

File: runtime/onert/core/include/backend/ITensor.h [hunk 2]
```diff
@@ -103,6 +103,17 @@ class ITensor
...
+  virtual bool is_subtensor() const { return false; }
+  virtual bool needMemoryMap() const { return false; }
+  virtual void enqueueWriteBuffer(const void *, bool)
+  {
+    throw std::runtime_error("This backend does not support enqueueWriteBuffer");
+  }
+  virtual void enqueueReadBuffer(void *, bool)
+  {
+    throw std::runtime_error("This backend does not support enqueueReadBuffer");
+  }
```
Comment (wateret) [inline runtime/onert/core/include/backend/ITensor.h:116]: Frankly I'm not a big fan of these introduced, but I also admit that there is another way that can be done quickly. We may accept this and we'll do some refactoring later.

**Review by wateret — DISMISSED** [review]
LGTM

**Review by hseok-oh — DISMISSED** [review]

File: runtime/onert/core/src/backend/controlflow/kernel/PermuteLayer.cc [hunk 3]
```diff
@@ -217,36 +230,54 @@ void PermuteLayer::run()
...
+            // This is more effective than multi-threading
+            src->access([&](backend::ITensor &) { dst->enqueueWriteBuffer(src->buffer(), false); });
+          }
+          else
+          {
+            _buffers_map[dst].reserve(dst->total_size());
+            auto dst_buffer = _buffers_map[dst].data();
+
+            src->access([&](backend::ITensor &) { runPermuteTasks(src, dst_buffer); });
+            dst->enqueueWriteBuffer(dst_buffer, false);
```
Comment (ragmani) [inline runtime/onert/core/src/backend/controlflow/kernel/PermuteLayer.cc:264]: This block can be slower than calling ` runPermuteTasks(src, dst->buffer());` if padding size is big.

**Review by hyunsik-yoon — APPROVED** [review]
LGTM

**Review by hseok-oh — APPROVED** [review]

**Comment by ragmani** [comment]
@wateret PTAL

**Review by wateret — APPROVED** [review]
LGTM

**Pull request merged**